Automattic/wp-calypso
# Calypso Build Package: Add webpack.config.js

#### Changes proposed in this Pull Request

See pafL3P-oe-p2:

> <h3>Problem: Webpack config re-use is suboptimal</h3>
>
> While we’re able to re-use parts of Calypso’s webpack config thanks to the shards concept (i.e. <a href="https://github.com/Automattic/wp-calypso/tree/924fb2c55b1237d13efacc5248bbc55bb6a0c910/packages/calypso-build/webpack">small files each containing specific parts of the config</a>), the resulting <code>webpack.config.js</code> files are still somewhat <a href="https://github.com/Automattic/wp-calypso/blob/924fb2c55b1237d13efacc5248bbc55bb6a0c910/webpack.config.js">verbose</a> and <a href="https://github.com/Automattic/jetpack/blob/31ce174fb25289c4c238b3f890d7fffbc2ef9c63/webpack.config.extensions.js">redundant</a>. The shards, as we are using them now, have never been the envisioned be-all, end-all, goal for an interface of a re-usable webpack config -- we even had a call with @blowery from +teamcalypsop2 on some options there (tl;dr: It’s complicated).
> 
> <h3>Solution: Parametrized `@automattic/calypso-build/webpack.config.js`</h3>
> 
> I think that as an intermediate step, it makes sense to add a parametrized <code>webpack.config.js</code> to <code>@automattic/calypso-build</code> that can then be less redundantly instantiated by consumers. The interface is still going to be suboptimal (essentially the union of all shards’ parameters), but IMO at least not worse than directly using the shards. Thanks to package versioning, we can release a new version featuring a better interface once we come up with one (and being used by a variety of different projects should help inform that interface).

The file used in here is essentially `webpack.config.extensions.js` minus the entry point magic, and minus `CopyWebpackPlugin` (about which `yarn` complained that it wasn't installed :roll_eyes: -- but we might be able to get rid of that anyway, see https://github.com/Automattic/jetpack/pull/11801).

#### Testing instructions

```
nvm use
cd packages/calypso-build
npm link
cd ../../
cd ../jetpack # Or wherever your local copy of Jetpack is
git checkout update/use-calypso-build-webpack-config
nvm use 10.15.2
yarn distclean && yarn
npm link @automattic/calypso-build
yarn build-extensions
```

Verify that blocks are built in your Jetpack directories `_inc/blocks/` folder.

#### Follow-up (this PR or separately)

Use this config file to build o2 blocks.

/cc @sirreal @simison
